text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,970 @@
+
+<!--
+Why are we doing this? What use cases does it support? What is the expected outcome?
+-->
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The `ops::ControlFlow` type
+
+This is a simple enum:
+```rust
+enum ControlFlow<B, C = ()> {
+    /// Exit the operation without running subsequent phases.
+    Break(B),
+    /// Move on to the next phase of the operation as normal.
+    Continue(C),
+}
+```
+
+It's intended for exposing things (like graph traversals or visitor) where you want the user to be able to choose whether to exit early.  Using an enum is clearer than just using a bool -- what did `false` mean again? -- as well as [allows it to carry a value](https://github.com/rust-lang/rust/pull/78779#pullrequestreview-524885131), if desired.
+
+For example, you could use it to expose a simple tree traversal in a way that lets the caller exit early if they want:
+```rust
+impl<T> TreeNode<T> {
+    fn traverse_inorder<B>(&self, mut f: impl FnMut(&T) -> ControlFlow<B>) -> ControlFlow<B> {
+        if let Some(left) = &self.left {
+            left.traverse_inorder(&mut f)?;
+        }
+        f(&self.value)?;
+        if let Some(right) = &self.right {
+            right.traverse_inorder(&mut f)?;
+        }
+        ControlFlow::Continue(())
+    }
+}
+```
+<!-- https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=286ac85c43e5bf242b5431b4f6f63386 -->
+
+
+## The `Try` trait
+
+The `ops::Try` trait describes a type's behavior when used with the `?` operator, like how the `ops::Add` trait describes its behavior when used with the `+` operator.
+
+At its core, the `?` operator is about splitting a type into its two parts:
+
+- The *output* that will be returned from the `?` expression, with which the program will continue, and
+- The *residual* that will be returned to the calling code, as an early exit from the normal flow.
+
+(Oxford's definition for a residual is "a quantity remaining after other things have been subtracted or allowed for", thus the use here.)
+
+The `Try` trait also has facilities for rebuilding a type from either of its parts.  This is needed to build the final return value from a function, both in `?` and in methods generic over multiple types implementing `Try`.
+
+Here's a quick overview of a few standard types which implement `Try`, their corresponding output and residual types, and the functions which convert between them.
+(Full details will come later; the goal for now is just to get the general idea.)
+
+```text
++-------------+                             +-------------------+                          +-------------------+
+| Try::Output |                             |      Try Type     |                          |   Try::Residual   |
```

Comment:
Are the lines inside the box labeled `Try::Residual` meant to be the definition of that associated type for the corresponding type in the middle box labelled `Try Type` ? That was my first impression...

... and that first impression matches up with the standard implementations. But then your implemenation for `ControlFlow` ends up returning a `ControlFlow<ControlFlow<..>, ..>` ... which I guess you expected, but seems counter-intuitive to me...

Reply:
That's correct.  All the things about needing the "colouring" of the residual apply for direct uses of `ControlFlow` too -- using `?` on a `ControlFlow<Result<!, E>>` in a `-> Result` method shouldn't work either.

So yes, the [nested `ControlFlow` in its `impl`](https://github.com/scottmcm/rust/blob/do-or-do-not-edition/library/core/src/ops/control_flow.rs#L101) is a bit odd, but I think it's correct.  For `Result` it's `-> ControlFlow<Result<!, E>, T>` and for `ControlFlow` it's `-> ControlFlow<ControlFlow<B, !>, C>`.

(And, for coherence, it's really nice that `FromResidual` is on `ControlFlow<B, !>`, not on uncovered `B`.  It's also a big help for the future possibilities, but of course those aren't directly controlling for this change.)